Allow inlining into trait constructors

lrytz opened the pull request:
Fixes scala/scala-dev#350

lrytz committed:
Allow inlining into trait constructors

Fixes scala-dev#350

retronym commented:
Thanks, this will eliminate a closure allocation in `trait AliasTypeRef { require(...) }`!

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/opt/InlinerHeuristics.scala:
```diff
@@ -64,7 +64,7 @@ class InlinerHeuristics[BT <: BTypes](val bTypes: BT) {
   }
 
-  private def isTraitStaticSuperAccessorName(s: String) = s.endsWith("$")
+  private def isTraitStaticSuperAccessorName(s: String) = s != "$init$" && s.endsWith("$")
```
Do we emit any other static methods (e.g. static-ified lambda lifted or lambda impl methods) that could possibly have the trailing `$` and also be misclassified?

retronym replied:
... it might be more robust to look for a static method named `foo$(self)` that only calls `self.foo`.

lrytz replied:
done, thanks

lrytz committed:
Clean up some testing in the backend

Build in the facility to keep certain per-run caches alive after
a run (to allow inspecting them). This was previously done by
hand, requiring every test in a file to clear those caches
before running a new test. Forgetting to do so would lead to
cross-talk.

SethTisue commented:
@lrytz should this go in 2.12.2?

lrytz commented:
It's not a regression and a benign bug, so 2.12.3 is just fine.

retronym commented:
@lrytz This needs a rebase (https://github.com/scala/scala/compare/2.12.x...retronym:review/5809?expand=1 should do the trick). Otherwise LGTM.

lrytz commented:
Thanks for the rebase! I looked at the diff of the diff, LGTM.
